From http://gcc.gnu.org/ml/gcc/2008-01/msg00306.html

This fix removes a particular warning "array subscript is above array
bounds" from being produced in some circumstances.  It is significant
because one of the circumstances comes up while building binutils
later on, in line 1388 of tc-mips.c.  Since the binutils build specifies
"-Werror", the warning causes the binutils build to abort.

The patch is by Richard Sandiford.

Index: gcc/global.c
===================================================================
--- gcc/global.c	2008-01-19 23:54:32.000000000 +0000
+++ gcc/global.c	2008-01-19 23:54:56.000000000 +0000
@@ -1016,8 +1016,13 @@ find_reg (int num, HARD_REG_SET losers, 
   if (allocno[num].no_eh_reg)
     {
       unsigned int j;
-      for (j = 0; EH_RETURN_DATA_REGNO (j) != INVALID_REGNUM; j++)
-	SET_HARD_REG_BIT (used1, EH_RETURN_DATA_REGNO (j));
+      for (j = 0; ; ++j)
+	{
+	  unsigned int regno = EH_RETURN_DATA_REGNO (j);
+	  if (regno == INVALID_REGNUM)
+	    break;
+	  SET_HARD_REG_BIT (used1, regno);
+	}
     }
 #endif
